src/evaluate/evaluator/question_answering.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2022 The HuggingFace Evaluate Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Dict, Optional, Tuple, Union
+
+# Lint as: python3
+from datasets import Dataset, load_dataset
+
+
+try:
+    from transformers import Pipeline, PreTrainedModel, PreTrainedTokenizer, TFPreTrainedModel
+
+    TRANSFORMERS_AVAILABLE = True
+except ImportError:
+    TRANSFORMERS_AVAILABLE = False
+
+from typing_extensions import Literal
+
+from ..module import EvaluationModule
+from ..utils.logging import get_logger
+from .base import Evaluator
+
+
+logger = get_logger(__name__)
+
+
+class QuestionAnsweringEvaluator(Evaluator):
+    """
+    Question answering evaluator. This evaluator handles
+    [**extractive** question answering](https://huggingface.co/docs/transformers/task_summary#extractive-question-answering),
+    where the answer to the question is extracted from a context.
+
+    This question answering evaluator can currently be loaded from [`evaluator`] using the default task name
+    `question-answering`.
+
+    Methods in this class assume a data format compatible with the
+    [`QuestionAnsweringPipeline`](https://huggingface.co/docs/transformers/en/main_classes/pipelines#transformers.QuestionAnsweringPipeline).
+    """
+
+    def __init__(self, task="question-answering", default_metric_name=None):
+        super().__init__(task, default_metric_name=default_metric_name)
+
+    def prepare_data(
+        self,
+        data: Union[str, Dataset],
+        question_column: str,
+        context_column: str,
+        id_column: str,
+        label_column: str,
+    ):
+        """Prepare data."""
+        if data is None:
+            raise ValueError(
+                "Please specify a valid `data` object - either a `str` with a name or a `Dataset` object."
+            )
+        data = load_dataset(data) if isinstance(data, str) else data
```

Comment:
Seeing this now I realize that this will probably fail when data is loaded via a string: dataset loaded that way come with splits ("train", "validation", "test") in a `DatasetDict`.

We probably need something like:
```py
splits = list(data.keys())
for split in ["test", "validation", "train"]:
if split in splits:
    data = data[split]
    break
```
cc @ola13

Reply:
Could we just have an argument `split`?

Reply:
i think it is ok to have a default behaviour like that, we can log which split was found and used. I would like to avoid too many args that are passed through. if a user wants a specific split they can always just load the dataset themselves and pass it directly to the evaluator.

Reply:
BTW we can fix this in a separate PR.

Reply:
Yes, let's do it in an other PR.